File: be/app/schemas/accommodation.py
from datetime import time
from typing import List, Optional

from pydantic import BaseModel

from app.schemas.base import BaseSchema, TimestampMixin


class AccommodationCategoryBase(BaseModel):
    name: str
    status: bool = True


class AccommodationCategoryCreate(AccommodationCategoryBase):
    pass


class AccommodationCategoryUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[bool] = None


class AccommodationCategory(AccommodationCategoryBase, TimestampMixin, BaseSchema):
    id: int


class AccommodationBase(BaseModel):
    user_id: int
    name: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    geom: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    category_id: int
    thumbnail_url: Optional[str] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    checkin_time: Optional[time] = None
    checkout_time: Optional[time] = None
    cancel_policy: Optional[str] = None
    pet_policy: Optional[str] = None
    child_policy: Optional[str] = None
    is_active: bool = True


class AccommodationCreate(AccommodationBase):
    pass


class AccommodationUpdate(BaseModel):
    name: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    geom: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    category_id: Optional[int] = None
    thumbnail_url: Optional[str] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    checkin_time: Optional[time] = None
    checkout_time: Optional[time] = None
    cancel_policy: Optional[str] = None
    pet_policy: Optional[str] = None
    child_policy: Optional[str] = None
    is_active: Optional[bool] = None
    popularity_score: Optional[float] = None


class Accommodation(AccommodationBase, TimestampMixin, BaseSchema):
    id: int
    popularity_score: float = 0


class AccommodationRoomBase(BaseModel):
    accommodation_id: int
    name: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    adult_capacity: int = 1
    child_capacity: int = 0
    room_area: Optional[int] = None
    bed_capacity: Optional[str] = None
    status: int = 1


class AccommodationRoomCreate(AccommodationRoomBase):
    pass


class AccommodationRoomUpdate(BaseModel):
    name: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    adult_capacity: Optional[int] = None
    child_capacity: Optional[int] = None
    room_area: Optional[int] = None
    bed_capacity: Optional[str] = None
    status: Optional[int] = None


class AccommodationRoom(AccommodationRoomBase, TimestampMixin, BaseSchema):
    id: int

File: java/src/main/java/com/williamtravel/app/entity/Accommodation.java
package com.williamtravel.app.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.Where;

import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "accommodations")
@Data
public class Accommodation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Integer id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "name", length = 255, nullable = false)
    private String name;

    @Column(name = "name_code", length = 255)
    private String nameCode;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "description_code", columnDefinition = "TEXT")
    private String descriptionCode;

    @Column(name = "latitude")
    private Double latitude;

    @Column(name = "longitude")
    private Double longitude;

    @Column(name = "geom", columnDefinition = "TEXT")
    private String geom;

    @Column(name = "address", length = 255)
    private String address;

    @Column(name = "city", length = 100)
    private String city;

    @ManyToOne
    @JoinColumn(name = "country_id")
    private Country country;

    @ManyToOne
    @JoinColumn(name = "region_id")
    private Region region;

    @ManyToOne
    @JoinColumn(name = "district_id")
    private District district;

    @ManyToOne
    @JoinColumn(name = "ward_id")
    private Ward ward;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private AccommodationCategory category;

    @Column(name = "thumbnail_url", length = 255)
    private String thumbnailUrl;

    @Column(name = "price_min")
    private Double priceMin;

    @Column(name = "price_max")
    private Double priceMax;

    @Column(name = "popularity_score")
    private Double popularityScore;

    @Column(name = "rating")
    private Double rating;

    @Column(name = "checkin_time")
    private LocalTime checkinTime;

    @Column(name = "checkout_time")
    private LocalTime checkoutTime;

    @Column(name = "cancel_policy", columnDefinition = "TEXT")
    private String cancelPolicy;

    @Column(name = "pet_policy", columnDefinition = "TEXT")
    private String petPolicy;

    @Column(name = "child_policy", columnDefinition = "TEXT")
    private String childPolicy;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    @OneToMany(mappedBy = "accommodation")
    private Set<AccommodationRoom> rooms = new HashSet<>();

    @OneToMany
    @JoinColumn(name = "reference_id")
    @Where(clause = "reference_type = 'accommodation'")
    private Set<Media> media = new HashSet<>();
}
